src/gretel_trainer/relational/extractor.py

```diff
@@ -0,0 +1,518 @@
+    of the total number of rows. A 0 value will just extracted headers and -1 will extract entire tables.
+
+    The default value, -1, implies that full tables should be extracted.
+    """
+
+    sample_mode: SampleMode = SampleMode.RANDOM
+
+    only: Optional[set[str]] = None
+    """
+    Only extract these tables. Cannot be used with `ignore.`
+    """
+
+    ignore: Optional[set[str]] = None
+    """
+    Ignore these tables during extraction. Cannot be used with `only.`
+    """
+
+    schema: str | None = None
+    """
+    Limit scope to a specific schema, this is a pass-through param to SQLAlchemy. It is not
+    supported by all dialects
+    """
+
+    def __post_init__(self):
+        errors = []
+
+        if self.sample_mode != SampleMode.RANDOM:
+            raise NotImplementedError(
+                "Additional sampling modes are not enabled yet, please use `random`"
+            )
+
+        if self.target_row_count < -1:
+            errors.append("The `target_row_count` must be -1 or higher")
+
+        if self.ignore is not None and self.only is not None:
+            errors.append("Cannot specify both `only` and `ignore` together")
+
+        if self.sample_mode not in ("random", "contiguous"):
+            errors.append("`sample_mode` must be one of 'random`, 'contiguous'")
+
+        if errors:
+            raise ValueError(f"The following errors occured: {', '.join(errors)}")
+
+    @property
+    def entire_table(self) -> bool:
+        return self.target_row_count == -1
+
+    @property
+    def empty_table(self) -> bool:
+        return self.target_row_count == 0
+
+    def should_skip_table(self, table_name: str) -> bool:
+        if self.only and table_name not in self.only:
+            return True
+
+        if self.ignore and table_name in self.ignore:
+            return True
+
+        return False
+
+
+def _determine_sample_size(config: ExtractorConfig, total_row_count: int) -> int:
+    """
+    Given the actual total row count of a table, determine how
+    many rows we should sample from it.
+    """
+    if config.target_row_count >= 1:
+        return int(config.target_row_count)
+
+    if config.entire_table:
+        return total_row_count
+
+    return int(total_row_count * config.target_row_count)
+
+
+@dataclass
+class _TableSession:
+    table: Table
+    engine: Engine
+
+    @property
+    def total_row_count(self) -> int:
+        with self.engine.connect() as conn:
+            query = select(func.count()).select_from(self.table)
+            count = conn.execute(query).scalar()
+            return 0 if count is None else int(count)
+
+    @property
+    def total_column_count(self) -> int:
+        return len(self.columns)
+
+    @property
+    def columns(self) -> list[str]:
+        return [column.name for column in self.table.columns]
+
+
+@dataclass
+class _PKValues:
+    """
+    Contains information that is needed to sample rows from an intermediate table
```

Comment:
Is there a reason you use "intermediate" throughout and not "parent"? Intermediate makes me think that it's maybe some temporary/ephemeral table that gets "passed through" on the way to some more terminal table.

Reply:
Can't a table be both a parent and a child though? Maybe it's better to just say non-leaf.

Reply:
Yeah... "non-leaf" might be least confusing here. FWIW though, you could describe the leaf tables as "non-parent" tables (because they don't have children i.e. tables with FKs pointing to them) that may or may not themselves be children. In other words, regardless of whether a table is a child or not, if it is a parent we'll be using _PKValues and if not we sample with the `SampleMode` enum.